src/DynamicRupture/Output/OutputAux.cpp

```diff
@@ -0,0 +1,276 @@
+    }
+  }
+  return std::make_pair(nearestPoint, shortestDistance);
+}
+
+void assignNearestGaussianPoints(ReceiverPointsT& geoPoints) {
+  std::shared_ptr<double[]> weights = nullptr;
+  std::shared_ptr<double[]> pointsData = nullptr;
+  unsigned numPoints{};
+
+  std::tie(numPoints, weights, pointsData) = generateTriangleQuadrature(CONVERGENCE_ORDER + 1);
+  double(*trianglePoints2D)[2] = reshape<2>(&pointsData[0]);
+
+  for (auto& geoPoint : geoPoints) {
+
+    double targetPoint2D[2];
+    transformations::XiEtaZeta2chiTau(
+        geoPoint.localFaceSideId, geoPoint.reference.coords, targetPoint2D);
+
+    int nearestPoint{-1};
+    double shortestDistance = std::numeric_limits<double>::max();
+    std::tie(nearestPoint, shortestDistance) =
+        getNearestFacePoint(targetPoint2D, trianglePoints2D, numPoints);
+    geoPoint.nearestGpIndex = nearestPoint;
+  }
+}
+
+int getClosestInternalStroudGp(int nearestGpIndex, int nPoly) {
+  int i1 = int((nearestGpIndex - 1) / (nPoly + 2)) + 1;
+  int j1 = (nearestGpIndex - 1) % (nPoly + 2) + 1;
+  if (i1 == 1) {
+    i1 = i1 + 1;
+  } else if (i1 == (nPoly + 2)) {
+    i1 = i1 - 1;
+  }
+
+  if (j1 == 1) {
+    j1 = j1 + 1;
+  } else if (j1 == (nPoly + 2)) {
+    j1 = j1 - 1;
+  }
+  return (i1 - 1) * (nPoly + 2) + j1;
+}
+
+void projectPointToFace(ExtVrtxCoords& point,
+                        const ExtTriangle& face,
+                        const VrtxCoords faceNormal) {
+  auto distance = getDistanceFromPointToFace(point, face, faceNormal);
+  double faceNormalLength = MeshTools::norm(faceNormal);
+  auto adjustedDistance = distance / faceNormalLength;
+
+  for (int i = 0; i < 3; ++i) {
+    point.coords[i] += adjustedDistance * faceNormal[i];
+  }
+}
+
+double getDistanceFromPointToFace(const ExtVrtxCoords& point,
+                                  const ExtTriangle& face,
+                                  const VrtxCoords faceNormal) {
+
+  VrtxCoords diff{0.0, 0.0, 0.0};
+  MeshTools::sub(face.p1.coords, point.coords, diff);
+
+  // Note: faceNormal may not be precisely a unit vector
+  double faceNormalLength = MeshTools::norm(faceNormal);
+  return MeshTools::dot(faceNormal, diff) / faceNormalLength;
+}
+
+PlusMinusBasisFunctionsT getPlusMinusBasisFunctions(const VrtxCoords pointCoords,
+                                                    const VrtxCoords* plusElementCoords[4],
+                                                    const VrtxCoords* minusElementCoords[4]) {
+
+  Eigen::Vector3d point(pointCoords[0], pointCoords[1], pointCoords[2]);
+
+  auto getBasisFunctions = [&point](const VrtxCoords* elementCoords[4]) {
+    auto referenceCoords = transformations::tetrahedronGlobalToReference(
+        *elementCoords[0], *elementCoords[1], *elementCoords[2], *elementCoords[3], point);
+    basisFunction::SampledBasisFunctions<real> sampler(
+        CONVERGENCE_ORDER, referenceCoords[0], referenceCoords[1], referenceCoords[2]);
+    return sampler.m_data;
```

Comment:
Yes, but a copy of `m_data` is going to be made because it is an instance of `std::vector`

```
template<class T>
class SampledBasisFunctions {
  static_assert(std::is_arithmetic<T>::value, "Type T for SampledBasisFunctions must be arithmetic.");
  
public:
    /** The basis function samples */
    std::vector<T> m_data;
```

So, the copy is going to be made before `SampledBasisFunctions`'s destructor is called.  Additionally a `return value optimization` is going take place